refactor integrated alerts algo and add dist alerts

solomon-negusse opened the pull request:
<!-- Please refer to our contributing documentation for any questions on submitting a pull request, or let us know here if you need any help -->

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [ ] Make sure you are requesting to pull a topic/feature/bugfix branch (right side). Don't request your master!
- [ ] Make sure you are making a pull request against the develop branch (left side). Also you should start your branch off our develop.
- [ ] Check the commit's or even all commits' message styles matches our requested structure.
- [ ] Check your code additions will fail neither code linting checks nor unit test.



## Pull request type

<!-- Please do not submit updates to dependencies unless it fixes an issue. --> 

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. --> 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 


## What is the new behavior?
<!-- Please describe the behavior or changes that are being added by this PR. -->

Adds algorithm to decode and render DIST alerts

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->

solomon-negusse committed:
refactor integrated alerts algo and add dist alerts

gtempus commented on app/routes/titiler/algorithms/alerts.py:
```diff
@@ -9,31 +9,29 @@
+
+    title: str = "Deforestation Alerts"
     description: str = "Decode and visualize alerts"
+    conf_colors: OrderedDict = OrderedDict(
+        {
+            DeforestationAlertConfidence.low: {
+                "confidence": 2,
+                "colors": (237, 164, 194),
+            },
+            DeforestationAlertConfidence.high: {
+                "confidence": 3,
+                "colors": (220, 102, 153),
+            },
+            DeforestationAlertConfidence.highest: {
+                "confidence": 4,
+                "colors": (201, 42, 109),
+            },
+        }
+    )
```
I ❤️ going with the immutable tuple here, @solomon-negusse.

I also like the descriptiveness of the _red, green, blue_ keys of the original `dict`. 

Did you consider using a namedTuple to get the best of both worlds?
Something like:
```python
# Define a namedtuple for Colors (immutable)
Colors = namedtuple('Colors', ['red', 'green', 'blue'])

# Define a namedtuple for DeforestationAlert (immutable)
DeforestationAlert = namedtuple('DeforestationAlert', ['confidence', 'colors'])

# Assuming DeforestationAlertConfidence is an Enum or similar object
conf_colors: OrderedDict = OrderedDict(
    {
        DeforestationAlertConfidence.low: DeforestationAlert(confidence=2, colors=Colors(237, 164, 194)),
        DeforestationAlertConfidence.high: DeforestationAlert(confidence=3, colors=Colors(220, 102, 153)),
        DeforestationAlertConfidence.highest: DeforestationAlert(confidence=4, colors=Colors(201, 42, 109)),
    }
)
```

solomon-negusse replied:
Great idea, thanks!

gtempus commented on app/routes/titiler/algorithms/alerts.py:
```diff
@@ -9,31 +9,29 @@
+    record_start_date: str = "2014-12-31"
 
     # Parameters
     default_start_date: str = (date.today() - relativedelta(days=180)).strftime(
```
This is a nitpick but I thought I'd mention it:
I noticed that `data.today()` is called a couple of times. Is it worth it to just call it once? 🤷

solomon-negusse replied:
Always good to DRY,  let me update it.

gtempus reviewed (commented):
I think you made a good design decision, considering there are only two types of `Alerts` in our TiTiler `Algorithm` domain right now. 

Regarding `__call__` and its current structure:
It's currently in a long imperative style. What about breaking it up into its principal parts? 

Namely: 
- assigning RGB values
- creating the mask
- and applying the mask. 

It would give it a more declarative feel, and tests could zero in on specific behavior at the unit level. 

I was going to try to write a few tests myself, but I ran out of time. 😦

solomon-negusse committed:
use namedtuple both for immutability and access with labels

solomon-negusse committed:
refactor to modularize and slim down __call__ method

solomon-negusse committed:
encapsulate alpha logic in a method

solomon-negusse committed:
add unit tests for alerts titiler algorithm

solomon-negusse commented:
> I think you made a good design decision, considering there are only two types of `Alerts` in our TiTiler `Algorithm` domain right now.
> 
> Regarding `__call__` and its current structure: It's currently in a long imperative style. What about breaking it up into its principal parts?
> 
> Namely:
> 
>     * assigning RGB values
> 
>     * creating the mask
> 
>     * and applying the mask.
> 
> 
> It would give it a more declarative feel, and tests could zero in on specific behavior at the unit level.
> 
> I was going to try to write a few tests myself, but I ran out of time. 😦

Thanks for the feedback, @gtempus.  I have refactored the `Alerts` class and added tests.

danscales commented on app/routes/titiler/algorithms/dist_alerts.py:
```diff
@@ -0,0 +1,27 @@
+        }
+    )
+
+    record_start_date: str = "2020-12-13"
```
Did you flip the digits?  The start date is "2020-12-31", according to the spec on VEG-DIST-DATE

solomon-negusse replied:
Ah, typo in our confluence document where I took this from then.

danscales commented on app/routes/titiler/algorithms/dist_alerts.py:
```diff
@@ -0,0 +1,27 @@
+
+
+class DISTAlerts(Alerts):
+    title: str = "Land Disturbunce (DIST) Alerts"
```
Disturbunce -> Disturbance

danscales commented on app/routes/titiler/algorithms/dist_alerts.py:
```diff
@@ -0,0 +1,27 @@
+
+    Colors: namedtuple = namedtuple("Colors", ["red", "green", "blue"])
+    DeforestationAlert: namedtuple = namedtuple(
+        "DeforestationAlert", ["confidence", "colors"]
```
Should this be "DisturbanceAlert" ?  DIST-Alerts covers more than forests?

solomon-negusse replied:
Since there's just a single definition of that in alerts.py now, I ended up calling it `AlertConfig` and also removed `deforestation` from a couple of variables that are common to both alert systems.

danscales commented on app/routes/titiler/algorithms/dist_alerts.py:
```diff
@@ -0,0 +1,27 @@
+    title: str = "Land Disturbunce (DIST) Alerts"
+    description: str = "Decode and visualize DIST alerts"
+
+    Colors: namedtuple = namedtuple("Colors", ["red", "green", "blue"])
```
Any need to include this if it is the same as the Base Class?

solomon-negusse replied:
Yes, they are not accessible to the subclasses. I ended up taking them out of the class and reusing them both in dist and integrated alerts modules now.

danscales commented on app/routes/titiler/algorithms/integrated_alerts.py:
```diff
@@ -0,0 +1,30 @@
+    title: str = "Integrated Deforestation Alerts"
+    description: str = "Decode and visualize deforestation alerts"
+
+    Colors: namedtuple = namedtuple("Colors", ["red", "green", "blue"])
```
Same here, any reason to include Colors here if it's the same as the base class?

solomon-negusse committed:
reuse alert config named tuple definitions; rename deforestationalert->alert to account for disturbance alerts

solomon-negusse committed:
fix start date

solomon-negusse committed:
fix typo

solomon-negusse committed:
fix typo

gtempus reviewed (approved):
This looks great @solomon-negusse! 💪 
It's well organized, there is a separation of concerns, and it's tested. 💥
